### PR TITLE
Move existing Filebeat generator logic to package `generator`

### PR DESCRIPTION
**Pull request by kvch**
I am splitting up https://github.com/elastic/beats/pull/6912 to make rebasing it a bit easier.
This PR moves the current generator logic, including module, fileset and fields.yml generators, to a different package.

Blocks https://github.com/elastic/beats/pull/6912

### CODE REVIEW

File: filebeat/generator/module/module.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+// Generate generates a new module.
+// If module exists, error is returned.
+func Generate(module, modulesPath, beatsPath string) error {
+	modulePath := path.Join(modulesPath, "module", module)
+	if generator.DirExists(modulePath) {
+		return fmt.Errorf("module already exists: %s", module)
+	}
+
+	err := generator.CreateDirectories(modulePath, []string{path.Join("_meta", "kibana", "6")})
```
[the remark this answers is not in the document]
Comment (kvch) [reply filebeat/generator/module/module.go]: Should we make it configurable? Maybe passing a param using a flag e.g. `--kibana-version`? It could be 6 by default until we release version 7.

[the remark this answers is not in the document]
Comment (ruflin) [reply filebeat/generator/module/module.go]: Instead of having the module generator creating the correct directory, it perhaps should be the kibana exporting command that takes care of the directory creation. So we don't create the directories here but they are only add when someone actually exports dashboards. It would be mean we have to add some knowledge about modules and Kibana directories to the exported but I think that would make sense.

[the remark this answers is not in the document]
Comment (kvch) [reply filebeat/generator/module/module.go]: I think it's a good idea.
I am wondering if there are use cases when a users would need directories without calling `export`. But I don't see any.

File: filebeat/generator/module/module.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	if err != nil {
+		return err
+	}
+
+	replace := map[string]string{"module": module}
+	templatesPath := path.Join(beatsPath, "scripts", "module")
+	filesToCopy := []string{
```
[the remark this answers is not in the document]
Comment (kvch) [reply filebeat/generator/module/module.go]: it calls `generator.CopyTemplates` which not only copies the files, but substitutes the template strings in copied files. I think apart from copying, there is not so much to share.

I could move `Copy` from the package `mage` to `libbeat/utils` or `libbeat/common/utils` and use it from there. It would be a bit bigger undertaking, as right now files are copied one-by-one and after each file is copied, the substitution is done. If I use this, I need to change the algorithm to substitute template strings after the file is copied. It would also mean that each file would need to be opened, read and written twice, once to copy it and once to substitute strings.

However, if I do so, I would rather refactor all generators and other possible places where files are copied. So I would rather address this in a follow up PR. WDYT?

[the remark this answers is not in the document]
Comment (ruflin) [reply filebeat/generator/module/module.go]: SGTM

**Comment by ruflin** [comment]
How does this PR conflict with https://github.com/elastic/beats/pull/7533 ?

**Comment by kvch** [comment]
I am moving the code I edit in the other PR to a different package in this one.

**Comment by ruflin** [comment]
I'm ok with moving the refactoring to an other PR and merge this one. The reason I think it does not fit well into https://github.com/elastic/beats/pull/7506 is because there a large part of the code is also moved around which will make it very hard to follow the diff. So my preferred option would be to have a follow up PR of this that takes care of the refactoring and then https://github.com/elastic/beats/pull/7506 is rebase on top of it.

https://github.com/elastic/beats/pull/7506 is the reason for me the refactoring becomes more important as the code is moved from dev only usage to production usage.

**Comment by kvch** [comment]
Works for me.

**Comment by ruflin** [comment]
Ups, my comment above was intended to be on https://github.com/elastic/beats/pull/7533. That is why I referenced this PR. Let's sync up offline on the order and cleanup of the PR's.

**Comment by kvch** [comment]
Closing in favour of a new PR.
